ENH: Add where argument to np.mean

sgasse opened the pull request:
Harmonize the signature of np.mean with np.sum by adding a where
argument, closes gh-15818.

sgasse committed:
ENH: Add where argument to several functions

Harmonize the signature of np.mean, np.var np.std, np.any, np.all,
and their respective nd.array methods with np.sum by adding a where
argument, see gh-15818.

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -140,13 +141,18 @@ def _clip(a, min=None, max=None, out=None, *, casting=None, **kwargs):
             um.clip, a, min, max, out=out, casting=casting, **kwargs)
 
-def _mean(a, axis=None, dtype=None, out=None, keepdims=False):
+def _mean(a, axis=None, dtype=None, out=None, keepdims=False, where=True):
```
```suggestion
def _mean(a, axis=None, dtype=None, out=None, keepdims=False, *, where=True):
```

sgasse replied:
Ah, this makes sure that `where` can only be set as keyword argument, not positional, right? Thanks for the hint!

eric-wieser replied:
Exactly. We haven't done it in many places elsewhere because we only recently required python versions supporting this syntax.

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -140,13 +141,18 @@ def _clip(a, min=None, max=None, out=None, *, casting=None, **kwargs):
+        logic_mask = where * nu.ones_like(arr, dtype=nt.bool_)
+        rcount = umr_sum(logic_mask, axis, nt.intp, None, keepdims)
+
+    if (asanyarray(rcount) == 0).all():
```
```suggestion
    if np.any(rcount == 0):
```

sgasse replied:
Makes sense :+1:

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -140,13 +141,18 @@ def _clip(a, min=None, max=None, out=None, *, casting=None, **kwargs):
+        rcount = _count_reduce_items(arr, axis)
+    else:
+        logic_mask = where * nu.ones_like(arr, dtype=nt.bool_)
+        rcount = umr_sum(logic_mask, axis, nt.intp, None, keepdims)
```
If you moved these 5 lines into ` _count_reduce_items`, then I think you'd get `where` for free in `np.var` too.

sgasse replied:
Should be possible, although I still have to figure out how to abstract this part in `_var`:
```
    # Compute degrees of freedom and make sure it is not negative.
    rcount = max([rcount - ddof, 0])
```

I will add a separate to the same PR later.

eric-wieser replied:
`np.maximum(rcount - ddof, 0)`

sgasse replied:
Thank you, I figured it out and thought 'Nice, this was easy' but then I was fiddling a little with broadcasting (`arrmean` needs to have fixed dimensions but `rcount` can cause wrong broadcasting behavior depending on where it is fixed-dim or not).

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -140,13 +141,18 @@ def _clip(a, min=None, max=None, out=None, *, casting=None, **kwargs):
+    if isinstance(where, bool):
+        rcount = _count_reduce_items(arr, axis)
+    else:
+        logic_mask = where * nu.ones_like(arr, dtype=nt.bool_)
```
This would be faster as `np.broadcast_to(where, arr.shape)`

sgasse replied:
Alright! Is it OK if I then add `import numpy as np` at the top of the file? I am just wondering because all other imports are very specifiy. I tried importing just `broadcast_to` from `numpy.lib.stride_tricks` but this leads to fails in imports (circular imports?).

eric-wieser replied:
You could import from stride_tricks locally within this function, which would break the cycle.

sgasse replied:
Alright thank you for the hint! By now I need also `np.maximum` and `np.any`, so I would keep it as is.

eric-wieser replied:
All those other functions are accessible under `nu` or some other import in this file.

sgasse replied:
:+1:  Took me a while to figure out where which function comes from, but now `np.maximum` is replaced with `um.maximum` and `np.any` with `umr_any` (keeping in mind the differing default argument for axis).

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -5495,6 +5495,33 @@ def test_mean_axis_error(self):
+                            np.array([6., 7., 8., 9.]))
+        
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter('always')
```
I think you can use `with pytest.warns(RuntimeWarning) as w` here

sgasse replied:
You're right, now it's more concise. Is there a special reason why in other tests, `warnings.catch_warnings` is used?

eric-wieser replied:
We didn't use to have a pytest dependency.

sgasse commented on numpy/core/_methods.py:
```diff
@@ -56,16 +59,23 @@ def _any(a, axis=None, dtype=None, out=None, keepdims=False):
+                        keepdims)
     return items
 
+
```
This extra blank line was unintended. I will remove it with possible changes after the next review.

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -56,16 +59,23 @@ def _any(a, axis=None, dtype=None, out=None, keepdims=False):
-        items *= arr.shape[mu.normalize_axis_index(ax, arr.ndim)]
+def _count_reduce_items(arr, axis, keepdims=False, where=True):
+    if isinstance(where, bool):
+        # no boolean mask given, calculate items according to axis
```
This doesn't handle `mask = False` correctly.

sgasse replied:
I first thought about leaving it out because setting `where=False` would basically mean to intentionally calculate the mean or variance of an empty slice. But you are right, I should return zero in this case. The update is in the next version.

anirudh2290 replied:
This comment confused me a bit, is this still required. What if `where=True` is provided when calling np.mean.

sgasse replied:
The default would be to have `where=True`, however in this earlier version, I did not handle `where=False` specifically, because I thought using `where=False` would not make sense in general. However to ensure correct behavior even for this case (`where=False` means attempting to calculate the mean of an empty slice), I added a branch.
Does this make it clearer, @anirudh2290 ?

anirudh2290 replied:
okay, thanks @sgasse ! i guess i got confused probably because I didnt first understand that "no boolean mask given" is same as "where=True has the same behavior as when no boolean mask is given". This also makes me think if this will cause a failure when where=None (probably a not so common use case).

eric-wieser replied:
I don't think where=None is legal anyway, but perhaps worth checking

sgasse replied:
@anirudh2290 `where=None` and `where=False` indeed lead to the divide-by-zero warning and `nan` as result. I am not sure this is how it should be, e.g. `arr.sum(where=None)` returns zero without any warning. Should I adapt `mean` in that way or would it clutter the function too much for catching an illegal input?

eric-wieser replied:
That behavior sounds correct to me. Mean is different to sum because it has to divide by the number of non-masked elements.

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -173,12 +184,13 @@ def _mean(a, axis=None, dtype=None, out=None, keepdims=False):
     arr = asanyarray(a)
 
-    rcount = _count_reduce_items(arr, axis)
+    rcount = _count_reduce_items(arr, axis, keepdims=keepdims, where=where)
```
Can you just pass `keepdims=True` here to save needing the reshape below?

sgasse replied:
I think not, let me try to explain:
- `arrmean` in line 204 has to be calculated with `keepdims=True`, otherwise the substraction from the array can be faulty.
- If the user specifies `keepdims=False` for `var`, `rcount` is thus calculated with it. The broadcasting of `arrmean` e.g. shape `(N, 1)` and `rcount` e.g. shape `(N, )` in line 208 at `true_divide` leads to an array of shape `(N, N)` that can no longer be stored in `out`.
- This means that the division in line 208 should definitely happen with a _shaped_ `rcount`.
- However if we would set `keepdims=True` in line 191 at the calculation of `rcount`, we would not be able to use `true_divide` with `ret` and `rcount` in line 238. Let's say `ret` is of shape `(N, )` because the user wants `keepdims=False` and `rcount` would be of shape `(N, 1)`, the result would be an array of shape `(N, N)` which could not be stored in the `out` target.

Yesterday, I was going back and forth with this 'dependency tree'. The proposed solution seemed the least intrusive and most efficient to me. However if you see a better solution given the above mentioned limitations, I am happy to try it out :smiley:

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -213,10 +228,10 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
 
     # Compute degrees of freedom and make sure it is not negative.
-    rcount = max([rcount - ddof, 0])
+    rcount = np.maximum(rcount - ddof, 0)
```
```suggestion
    rcount = um.maximum(rcount - ddof, 0)
```

sgasse replied:
Thank you, done.

eric-wieser commented:
Thanks for the `var` support, this looks surprisingly straightforward.

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -56,14 +58,23 @@ def _any(a, axis=None, dtype=None, out=None, keepdims=False):
+        # no boolean mask given, calculate items according to axis
+        if axis is None:
+            axis = tuple(range(arr.ndim))
+        if not isinstance(axis, tuple):
```
```suggestion
        elif not isinstance(axis, tuple):
```

sgasse replied:
Good point!

sgasse commented on numpy/core/fromnumeric.py:
```diff
@@ -3284,6 +3286,9 @@ def mean(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
 
+    where : array_like of bool, optional
+        Elements to include in the mean. See `~numpy.ufunc.reduce` for details.
+
```
Does this have to be separated with a `.. versionadded:: <whatever>`?

eric-wieser replied:
That should be as the last line of this description

sgasse replied:
Alright, updated it to `1.19.0`, or does it have to be the next, so `1.19.1`?

eric-wieser replied:
1.19.0 is fine

sgasse commented on numpy/core/fromnumeric.py:
```diff
@@ -3522,6 +3530,10 @@ def var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=np._NoValue):
+    where : array_like of bool, optional
+        Elements to include in the variance. See `~numpy.ufunc.reduce` for
+        details.
+
```
Same as above, do we need a `.. versionadded:: <whatever>` here?

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -56,14 +56,25 @@ def _any(a, axis=None, dtype=None, out=None, keepdims=False):
-    for ax in axis:
-        items *= arr.shape[mu.normalize_axis_index(ax, arr.ndim)]
+def _count_reduce_items(arr, axis, keepdims=False, where=True):
+    if isinstance(where, bool):
```
We should check which cases `ufunc.reduce` (c function name is something like GenericReduction) has this optimization for, and try the same ones. In particular, I'm trying to remember if we optimized for `np.bool_`.

sgasse replied:
I found the `PyArray_GenericReduceFunction` in `numpy/core/src/multiarray/number.c` however that function looks more like a wrapper, being only 15 lines long. Could you point me in the right direction?

anirudh2290 replied:
I think @eric-wieser is indicating at this function : PyUFunc_GenericReduction in numpy/core/src/umath/ufunc_object.c . I wasn't able to find any special casing for np.bool except that where = True is being treated as if no where is provided.

sgasse replied:
OK thank you for checking @anirudh2290.

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -56,14 +56,25 @@ def _any(a, axis=None, dtype=None, out=None, keepdims=False):
+        for ax in axis:
+            items *= arr.shape[mu.normalize_axis_index(ax, arr.ndim)]
+    else:
+        # guarded to protect circular imports
```
Worth a TODO here about this sum bring more expensive than needed for when where is broadcast along a non-reduction axis.

sgasse replied:
Good point, added a TODO note.

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -140,13 +151,13 @@ def _clip(a, min=None, max=None, out=None, *, casting=None, **kwargs):
-    if rcount == 0:
+
+    rcount = _count_reduce_items(arr, axis, keepdims=keepdims, where=where)
+    if umr_any(rcount == 0, axis=None):
```
Isn't none the default?

eric-wieser replied:
Nevermind, see you mention that above.

sgasse replied:
This is what I thought, because for `np.any`, the default is `axis=None`. However when the checks failed, I investigated it and found that for `numpy.ufunc.reduce`, the default is `axis=0`
[Version 1.19.0 `numpy.ufunc.reduce` doc](https://numpy.org/devdocs/reference/generated/numpy.ufunc.reduce.html)

sgasse replied:
I don't know if this difference is intentional. Maybe it would make sense to harmonize this in another issue in the future to increase the consistency.

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -213,10 +228,10 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
         x = um.multiply(x, um.conjugate(x), out=x).real
 
-    ret = umr_sum(x, axis, dtype, out, keepdims)
+    ret = umr_sum(x, axis, dtype, out, keepdims=keepdims, where=where)
```
My recommendation would be to add
```
    if not keepdims:
        rcount = rcount.squeeze(axis=axis)
```
here

eric-wieser replied:
Ugh, I see this doesn't work either.

sgasse replied:
So should I maybe leave it as is?

eric-wieser replied:
Probably best

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -189,10 +201,13 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
+    arrmean = umr_sum(arr, axis, dtype, keepdims=True, where=where)
+    # The shape of rcount has to match arrmean to not change the shape of out
+    # in broadcasting. Otherwise, it cannot be stored back to arrmean.
+    div = rcount if isinstance(rcount, int) else rcount.reshape(arrmean.shape)
```
Which would let you eliminate this line

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -56,14 +56,25 @@ def _any(a, axis=None, dtype=None, out=None, keepdims=False):
+            axis = (axis,)
+        items = 1
+        for ax in axis:
+            items *= arr.shape[mu.normalize_axis_index(ax, arr.ndim)]
```
For consistency, I'd recommend:
```suggestion
            items *= arr.shape[mu.normalize_axis_index(ax, arr.ndim)]
        items = np.intp(items)
```

sgasse replied:
Thanks, good idea! I now initialize `items = nt.intp(1)` above to save the line and also return `nt.intp(0)` in the scenario above.

eric-wieser replied:
While saving a line, it's possible it's very marginally slower. Unlikely to be an issue though.

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -189,10 +204,14 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
+    arrmean = umr_sum(arr, axis, dtype, keepdims=True, where=where)
+    # The shape of rcount has to match arrmean to not change the shape of out
+    # in broadcasting. Otherwise, it cannot be stored back to arrmean.
+    div = (rcount if isinstance(rcount, nt.intp)
```
```suggestion
    div = (rcount if rcount.ndim == 0
```

sgasse replied:
:+1: nice, then it also fits on one line.

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -56,14 +56,28 @@ def _any(a, axis=None, dtype=None, out=None, keepdims=False):
-    for ax in axis:
-        items *= arr.shape[mu.normalize_axis_index(ax, arr.ndim)]
+def _count_reduce_items(arr, axis, keepdims=False, where=True):
+    if isinstance(where, bool):
```
```suggestion
    # fast-path for the default case
    if where is True:
```

sgasse replied:
Good suggestion, I will included it in the change.
However what should the expected result of `where=None` and `where=False` be? Right now, I do not catch those cases separately, so they go through the broadcast branch. This could be more efficient with something like this:
```
if where is None or where is False:
    # no elements in slice
    return nt.intp(0
```
but of course if this is not _legal_ case, it might not be needed.

If you run `arr.sum(where=False)` or `arr.sum(where=None)`, the result is always 0. However due to the division in `mean`, it turns out to be `nan` with a warning. Should this be caught and zero returned or is the warning alright since `False` and `None` are not among the expected inputs to `where`?

eric-wieser replied:
> Right now, I do not catch those cases separately, so they go through the broadcast branch.

The best way to make sure your corner cases are readonable is to not have to special case them at all.

> This could be more efficient with 

Remember, the reason we care about optimizing `where=True` is because it's the default, and the path taken by all existing code using `mean` today. We don't have any reason to care about optimizing other corner cases, all that matters is that we get them correct.

eric-wieser replied:
Also, `where=None` is only legal for the same reason as `where=""` or `where=MyFalsyObject()`.

sgasse replied:
> The best way to make sure your corner cases are reasonable is to not have to special case them at all. Sending them through the broadcast branch is the right thing to do (even if your optimization was correct).
> Remember, the reason we care about optimizing `where=True` is because it's the default, and the path taken by all existing code using `mean` today. We don't have any reason to care about optimizing other corner cases, all that matters is that we get them correct.

OK makes sense, thank you for clarifying!

eric-wieser commented on numpy/core/fromnumeric.py:
```diff
@@ -3237,12 +3237,14 @@ def around(a, decimals=0, out=None):
 @array_function_dispatch(_mean_dispatcher)
-def mean(a, axis=None, dtype=None, out=None, keepdims=np._NoValue):
+def mean(a, axis=None, dtype=None, out=None, keepdims=np._NoValue,
+         where=np._NoValue):
```
Let's make this keyword-only?
```suggestion
         *, where=np._NoValue):
```

sgasse replied:
Thanks, good suggestion! I also made the `where` argument keyword-only in `var(...)`.

eric-wieser commented on numpy/core/fromnumeric.py:
```diff
@@ -3237,12 +3237,14 @@ def around(a, decimals=0, out=None):
 
-def _mean_dispatcher(a, axis=None, dtype=None, out=None, keepdims=None):
+def _mean_dispatcher(a, axis=None, dtype=None, out=None, keepdims=None,
+                     where=None):
```
```suggestion
                     *, where=None):
```

sgasse replied:
Done, also for `var`.

eric-wieser commented on numpy/core/fromnumeric.py:
```diff
@@ -3237,12 +3237,14 @@ def around(a, decimals=0, out=None):
-def _mean_dispatcher(a, axis=None, dtype=None, out=None, keepdims=None):
+def _mean_dispatcher(a, axis=None, dtype=None, out=None, keepdims=None,
+                     where=None):
     return (a, out)
```
@mhvk, do you remember if reductions visit `type(where).__array_ufunc__`? If so, we should include `where` here too.

mhvk replied:
Right now, it is a bit of a mess: if `where` is a positional argument, then, yes, it is checked, while if it is a keyword argument, it is not (see #8892, second item).  At the time I noticed this, there was a bit of thought of whether one should generally accept overrides (even on, say, `axis`). I think that is still up in the air, but I think for `__array_function__` at least the idea that any array argument can override makes sense - and I'd rather adjust `__array_ufunc__` to do the same (e.g., come form of calculated mask).

eric-wieser replied:
```suggestion
    return (a, where, out)
```

sgasse replied:
@eric-wieser @mhvk thank you for the input. In this case, I must admit that I don't really understand how the dispatcher comes into play and what the effect of having `where` in the return of the dispatcher is. Could you point me to the right part of the documentation or explain it a little? I did integrate your suggested change in the commit. However just for `mean` so far, since I don't fully understand it yet. Should I also include it for `var`? (line `3486` in `fromnumeric.py`)

eric-wieser replied:
Yes, please do include it for `var`.

sgasse replied:
Done

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -189,10 +202,13 @@ def _var(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
+    arrmean = umr_sum(arr, axis, dtype, keepdims=True, where=where)
+    # The shape of rcount has to match arrmean to not change the shape of out
+    # in broadcasting. Otherwise, it cannot be stored back to arrmean.
+    div = (rcount if rcount.ndim == 0 else rcount.reshape(arrmean.shape))
```
I think this would be clearer as a regular `if` statement, with a comment for each branch.

In particular, I think it's worth drawing attention to the fact that `rcount.ndim == 0` is the optimized case from `_count_reduce_items`

sgasse replied:
Good suggestion, I will update it accordingly.

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -5499,6 +5499,28 @@ def test_mean_axis_error(self):
+                               [True],
+                               [True],
+                               [False]])
+        assert_allclose(a.mean(axis=1, where=True),
```
A test of `where=False` for mean and var would be good too.

sgasse replied:
Alright, I've added `where=False` tests for `mean`, `var` and `std`, all returning `nan` and raising a `RuntimeWarning`.

charris commented:
Needs a release note before it goes in.

rgommers reviewed (commented):
This looks like a useful improvement, thanks @sgasse for working on this. A couple of comments:
- Could you please add tests for both function and method (e.g. `np.mean` and `ndarray.mean`)?
- It would be useful to do all of these functions in `_methods.py` in one go (e.g. `std`, `any`). Or if that's too much, open an issue for the ones that still miss the `where` keyword. There's also other, logically similar ones like `np.median` defined in different `.py` files.

eric-wieser commented:
I suspect `any` and `all` quite deliberately missing the where argument, cc @mhvk.

`std` is a good spot, and an easy one too.

mhvk commented:
@eric-wieser - I fear it is much simpler: when I introduced `where` for the reductions, I simply didn't think about these ones - remembering #12644 and preceding attempts, I think much of the discussion was about how to get the C code make sense. I may have missed `any` and `all` simply because I searched for `initial` to add to and they didn't have that argument... Anyway, I would agree with @rgommers that one might as well include them now.

p.s. @sgasse, very happy to see this improvement!

sgasse commented:
@rgommers I'm glad you like it :smiley: Adding tests for both `np.mean` and `ndarray.mean` (and `var` etc.) make sense, I will look into it.

@eric-wieser @mhvk You're right, adding the where keyword to `std`, `any` and `all` should be easy. However to make comprehensive tests, I would look into it on the weekend.

sgasse commented:
Now, I've combined the changes to `mean` and `var` in one commit with the other changes.

- @rgommers I've updated the tests to check both the numpy function and the array method. Maybe you can have a look whether it's OK.
- @eric-wieser @mhvk With this commit, I also added the `where` argument to `np.std`, `np.any` and `np.all` including their array methods. Could you check whether it matches the overall structure? E.g. `np.prod` etc. also accept `where` as positional argument. However for `any` and `all`, I added it as keyword-only in line with `mean`, `var` and `std`.
- @charris could you please point me to where/how I add a release note?

There is also one slightly weird behavior in `np.any` and `np.all`. If the broadcasting of the `where` argument leads to checking empty slices, the truth value of the remaining dimensions get's broadcast in the result. However I guess it's OK regarding the input dimensions:
```
>>> a = np.array([[True, False, True],
...               [False, False, False],
...               [True, True, True]])
>>> wh_lower = np.array([[False],
...                      [False],
...                      [True]])
>>> a.all(where=wh_lower)
True # OK
>>> a.all(axis=0, where=wh_lower)
array([ True,  True,  True]) # OK
>>> a.all(axis=1, where=wh_lower)
array([ True,  True,  True]) # slightly strange since only one slice along 0 remains
```

eric-wieser commented:
> the truth value of the remaining dimensions get's broadcast in the result

I think you're misdiagnosing that. `all` on an empty slice is always true, and `any` on an empty slice is always false.

eric-wieser commented:
In principle, `_ptp` a few lines below could quite get a `where` argument too. I think we'd forward `where` to min and max, but not to subtract.

@mvhk, there's possibly an issue here - `where` for regular ufuncs means "write only these outputs", but for reductions it means "read only these inputs".

Do we want to allow specifying both somehow? This manifests itself in `ptp`, where we can choose between passing `where` to one or both of `subtract` and `min`.

mhvk commented:
@eric-wieser - conceptually, I prefer to think of `where` as "places where a calculation is done" even if in the regular ufuncs it is implemented as doing the calculation but just not writing out the result. With this in mind, it seems by most logical to have `where` apply to the first step, ie, pass on the `min` and `max`, but not to subtract.

eric-wieser commented:
@mhvk: I'm thinking of how to handle code like:
```python
a = np.add(b, c, where=mask)
a = np.subtract(b, c, where=~mask, out=a)
```
But where the functions in question are reductions

mhvk commented:
For the "output where", given that calculations are being done anyway, the only real case for having it as an alternative to just selecting after the fact would be to save memory. But with reductions, I think that case is much less strong, since the memory for the output will presumably already be much less than that for the input anyway. I.e., for your example, I think there would be little reason to deviate from the simple
```
a = np.min(b)
a[~mask] = np.max(b)
```

Aside: as I mentioned, the distinction between input and output where for regular ufunc calls would seem only a matter of implementation. Where it would matter is for `accumulate` - I'm not sure what the most logical behaviour would be:
1. Skip input, but write output. Effectively the same as replacing the input with `initial`.
2. Use input, but skip output. Why would that ever be useful?
3. Skip input, skip output. Bit weird to get those unfilled outputs.

Overall, thinking of it as a selection of inputs only seems to me the easiest.

seberg commented:
Its also relevant for gufuncs (of which reductions are a special case after all). There things are even more complicated, since `initial` is suddenly an array (i.e. say it could be the identity matrix).

I don't have an opinion on the where argument. Of that list above 1. seems generally pretty useful? Although, you still have the question of what the "element" is, e.g. the array element, or the inner-axes of the gufunc (the identity matrix above is arguably the element)?

For reductions, the elements are typically the normal array elements. So I wonder if the `out_where` probably ends up as way to speed up the case of
```
out_where == where.all(axis, keepdims=True)
```
In both cases you would want to write out `default` (unless you prefer to write out nothing at all), although in one case not having `default` may lead to an error, while in the other case it is clear that it cannot.

sgasse commented:
> I think you're misdiagnosing that. all on an empty slice is always true, and any on an empty slice is always false.

Ah, thank you @eric-wieser for clarifying that, I did not have it in mind.

> In principle, _ptp a few lines below could quite get a where argument too. I think we'd forward where to min and max, but not to subtract.

I gave it a go in my last commit, but it does not seem as easy as I thought. The `maximum.reduce` and `minimum.reduce` ufuncs both need an `initial` value when `where` is given. If `axis=None`, this is OK as any value of the array will serve that purpose. However, if `axis` and `where` are given at the same time, the calculation often returns a wrong result because `initial` would need to be set per reduction axis. Example:
```
>>> a = np.arange(9).reshape((3, 3))
>>> a
array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])
>>> wh_full = np.array([[True, True, False],
...                     [False, True, True],
...                     [True, False, True]])
>>> a.ptp(axis=1, where=wh_full)
array([1, 5, 8]) # should be [1, 1, 2]
```
@seberg you mentioned `gufuncs` that can take an array as `initial` value. Could this be a solution in this situation? 

In the last commit, I added a `RuntimeWarning` and a test, but the solution is of course not good and I am aware that it cannot stay that way. If someone has a suggestion, I'd be happy to add that.

sgasse commented:
> I gave it a go in my last commit, but it does not seem as easy as I thought. The maximum.reduce and minimum.reduce ufuncs both need an initial value when where is given. If axis=None, this is OK as any value of the array will serve that purpose. However, if axis and where are given at the same time, the calculation often returns a wrong result because initial would need to be set per reduction axis.

Maybe I found a solution that seems pretty obvious in hindsight. In the last commit, I first determine the datatype, get the machine limits and then set the machine minimum as initial value for the maximum calculation and vice versa. Don't know though whether this has adverse implications for performance. It is only evaluated when `where` is not `True`.

Tests for this are added. However, there is still one case where I am not sure if I see expected behavior. When I calculate `ptp` along an axis over an empty slice, I get `1` as result:
```
>>> a = np.arange(9).reshape((3, 3))
>>> a
array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])
>>> wh_partial = np.array([[True],
...                        [True],
...                        [False]])
>>> a.ptp(axis=1, where=wh_partial)
array([2, 2, 1]) # what should the last entry have been?
```

On a side-note: I ran all checks successfully locally. However after rebasing on master, I get two failed tests locally (`numpy/random/tests/test_generator_mt19937.py::TestIntegers::test_repeatability_32bit_boundary[4294967296-expected1]` and `numpy/random/tests/test_generator_mt19937.py::TestIntegers::test_repeatability_32bit_boundary_broadcasting`). Since I also get them when running tests on master without my patch, I still pushed and will later look into what could be wrong in my local setup.

eric-wieser commented on numpy/core/_methods.py:
```diff
@@ -242,10 +264,27 @@ def _std(a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
 
     return ret
 
-def _ptp(a, axis=None, out=None, keepdims=False):
+def _ptp(a, axis=None, out=None, keepdims=False, *, where=True):
+    max_call = (a, axis, None, out, keepdims)
+    min_call = (a, axis, None, None, keepdims)
+
+    if where is not True:
+        # guarded to protect circular imports
+        from numpy.core.getlimits import iinfo, finfo
+
+        if nt._kind_name(a.dtype) in ['uint', 'int']:
+            _info = iinfo(a.dtype)
+        elif nt._kind_name(a.dtype) in ['float', 'complex']:
+            _info = finfo(a.dtype)
+        else:
+            raise TypeError(('Cannot evaluate peak-to-peak for dtypes other '
+                             'than uint, int, float or complex'))
+        max_call += (_info.min, where)
+        min_call += (_info.max, where)
+
     return um.subtract(
-        umr_maximum(a, axis, None, out, keepdims),
-        umr_minimum(a, axis, None, None, keepdims),
+        umr_maximum(*max_call),
+        umr_minimum(*min_call),
         out
     )
```
This doesn't do the right thing for arrays containing `np.inf`. I recommend we skip `ptp` for now, and attempt it in a future PR - it's sufficiently different from `mean` and `var` that I think it's worth handling separately, so that we can get the rest merged.

sgasse replied:
Alright, I will remove changes made to `ptp` and upload a commit without it.

sgasse commented:
@eric-wieser I removed all changes related to `ptp`. Please let me know if other changes are required before the commit can be merged. Looking at the failed checks, it appears to me that they are unrelated to my changes. However I build locally on Linux and cannot test a build on Windows.

seberg commented on numpy/core/_methods.py:
```diff
@@ -50,20 +50,32 @@ def _prod(a, axis=None, dtype=None, out=None, keepdims=False,
-    for ax in axis:
-        items *= arr.shape[mu.normalize_axis_index(ax, arr.ndim)]
+def _any(a, axis=None, dtype=None, out=None, keepdims=False, *, where=True):
+    return umr_any(a, axis, dtype, out, keepdims, where=where)
```
The kwarg will have some performance implications for small arrays <1000 elements or so, until my parsing improvements would be merged. But just a note, to be honest there is not much we can do about it. It seems passing by position is thwarted by the `initial` argument (though that could possibly be changed).

@eric-wieser do you want to persue this for 1.19? I will push it off for now.

sgasse commented:
@eric-wieser would you have time to take a last quick glance at this so it could maybe go in?

seberg reviewed (commented):
Hmm, sorry for not moving this forward earlier.  @eric-wieser you were happy with this? It seems good to me, would do another more careful path before merging maybe, but seems all good.

@sgasse would you be willing to add an improvement release note for this?  since these are all reduction like functions, adding an argument that all reductions accept seems like a straight forward API decision.

sgasse commented:
Hey @seberg , nice to hear back on this pull request :)
Sure, later/tomorrow, I can update all the `versionadded` tags in my changes. Where exactly can I add the improvement release note?

seberg commented:
Thanks, releaes notes would go into `doc/release/upcoming_changes`, there is a readme in that file detailing the process.

sgasse commented:
Alright, thank you @seberg for the location of of the release notes. I added a document there and updated all the `versionadded` tags. Maybe you can have a look at the release note document.

seberg commented:
Sorry, I was AFK for a bit. The release notes seem long but fine.  @eric-wieser if there was anything you felt still doing here please have a quick look. I somewhat remembered that you almost had signed this off. But maybe that was not the case.

As a general note, the slowdown is:
```
In [2]: %timeit np.mean(arr)                                                                                            
10.8 µs ± 575 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
vs.
```
In [3]: %timeit np.mean(arr)                                                                                            
17.7 µs ± 132 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
For `arr.any()` the difference is 2.4µs vs. 3µs after.  Since this adds a reduce argument to reduce-like operations. I am generally in favor of adding this, but it is a fairly large speed impact for small arrays there with `mean`, and I am uncertain that it will be easy to micro-optimize away.

charris commented:
Ping everyone here. Sounds like this is ready.

seberg commented:
I am OK with putting it in as is. The only thing that made me pause in the first place was the slowdown, which is a silly reason probably.

sgasse commented:
@charris @seberg nice to hear back - should I rebase this or is it fine as is? I see the next major version is still 1.20, so this does not need to be updated, right?

charris commented:
In it goes, thanks @sgasse .
